File: Task Tracker Java/contest_java/src/main/java/com/example/contest/repo/participants/ParticipantsDBRepository.java
package com.example.contest.repo.participants;

import com.example.contest.domain.Participant;
import com.example.contest.domain.enums.AgeGroup;
import com.example.contest.domain.enums.Type;
import com.example.contest.utils.JDBCutils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;


import java.io.IOException;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.Period;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class ParticipantsDBRepository implements ParticipantsRepository {
    private JDBCutils jdbCutils;
    private static final Logger logger = LogManager.getLogger();

    public ParticipantsDBRepository(Properties props) {
        logger.info("Initializing Repository with {}", props);
        jdbCutils = new JDBCutils(props);
    }

    @Override
    public List<Participant> getAll() throws IOException {
        logger.traceEntry();
        List<Participant> participants = new ArrayList<>();

        String query = "SELECT * FROM participants";
        Connection connection = jdbCutils.getConnection();

        try (PreparedStatement preparedStatement = connection.prepareStatement(query)) {
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    int participantID = resultSet.getInt("participantid");
                    String name = resultSet.getString("name");
                    int age = resultSet.getInt("age");
                    Participant participant = new Participant(participantID, name, age);
                    participants.add(participant);
                }
            }
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println("DB Error " + ex);
            throw new RuntimeException(ex);
        }

        return participants;
    }

    @Override
    public boolean delete(Participant entity) throws IOException {
        logger.traceEntry("deleting participant {}", entity);

        String query = "DELETE FROM participants WHERE participantid=?";
        Connection connection = jdbCutils.getConnection();

        try (PreparedStatement preparedStatement = connection.prepareStatement(query)) {
            preparedStatement.setInt(1, entity.getId());
            int result = preparedStatement.executeUpdate();
            logger.traceExit("deleted {}", entity);
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println("DB Error " + ex);
            throw new RuntimeException(ex);
        }

        return false;
    }

    @Override
    public Participant update(Participant entity) throws IOException {
        logger.traceEntry("updating participant ");

        String query = "UPDATE participants SET name=?, age=? WHERE participantid=?";
        Connection connection = jdbCutils.getConnection();

        try (PreparedStatement preparedStatement = connection.prepareStatement(query)) {
            preparedStatement.setString(1, entity.getName());
            preparedStatement.setInt(2, entity.getAge());
            preparedStatement.setInt(3, entity.getId());
            int result = preparedStatement.executeUpdate();
            logger.traceExit("updated participant {}", entity);
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println("DB Error " + ex);
            throw new RuntimeException(ex);
        }

        return entity;
    }

    @Override
    public Participant save(Participant entity) throws IOException {
        logger.traceEntry("saving participant {}", entity);

        String query = "INSERT INTO participants(name, age) VALUES (?, ?)";
        Connection connection = jdbCutils.getConnection();

        try (PreparedStatement preparedStatement = connection.prepareStatement(query)) {
            preparedStatement.setString(1, entity.getName());
            preparedStatement.setInt(2, entity.getAge());
            int result = preparedStatement.executeUpdate();
            logger.trace("Saved {}", result);
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println("DB Error " + ex);
            throw new RuntimeException(ex);
        }

        return entity;
    }

    @Override
    public int size() {
        logger.traceEntry("counting participants");

        String query = "SELECT COUNT(*)  FROM participants";
        Connection connection = jdbCutils.getConnection();

        try (PreparedStatement preparedStatement = connection.prepareStatement(query)) {
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                if (resultSet.next()) {
                    logger.traceExit(resultSet.getInt("count(*)"));
                    return resultSet.getInt("count(*)");
                }
            }
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println("DB Error " + ex);
            throw new RuntimeException(ex);
        }

        return 0;
    }

    @Override
    public Participant findByID(int id) {
        String query = "SELECT * FROM participants WHERE participantid=?";
        Connection connection = jdbCutils.getConnection();

        try (PreparedStatement preparedStatement = connection.prepareStatement(query)) {
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                if (resultSet.next()) {
                    int participantID = resultSet.getInt("participantid");
                    String name = resultSet.getString("name");
                    int age = resultSet.getInt("age");
                    return new Participant(participantID, name, age);
                }
            }
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println("DB Error " + ex);
            throw new RuntimeException(ex);
        }

        return null;
    }

    @Override
    public List<Participant> filterByAgeType(Type type, AgeGroup ageGroup) {
        List<Participant> filteredParticipants = new ArrayList<>();
        String query = """
                select p.participantid, p.name, p.age from participants p  INNER JOIN participantstasks pt on p.participantid=pt.participantID
                                                      INNER JOIN tasks t on pt.taskID1=t.taskid OR pt.taskID2 = t.taskid
                where t.type=? AND t.agegroup=?""";
        Connection connection = jdbCutils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(query)) {
            preparedStatement.setString(1, String.valueOf(type));
            preparedStatement.setString(2, String.valueOf(ageGroup));
            ResultSet resultSet = preparedStatement.executeQuery();
            while (resultSet.next()){
                int participantID = resultSet.getInt("participantid");
                String name = resultSet.getString("name");
                int age = resultSet.getInt("age");
                Participant participant = new Participant(participantID, name, age);
                filteredParticipants.add(participant);
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
        return filteredParticipants;
    }
}

File: Task Tracker C#/contest/contestRepoDB/participants/ParticipantsDBRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using contestDomain;
using contestDomain.enums;
using Type = contestDomain.enums.Type;
using log4net;

namespace contestRepoDB.participants
{
    public class ParticipantsDBRepository : ParticipantsRepo
    {
        private static readonly ILog logger = LogManager.GetLogger("ParticipantsDBRepository");

        private IDictionary<string, string> Properties;

        public ParticipantsDBRepository(IDictionary<string, string> properties)
        {
            logger.Info("Creating Participants DB Repo");
            Properties = properties;
        }

        public Participant FindById(int id)
        {
            return null;
        }

        public List<Participant> FindParticipantsByAgeAndType(Type type, AgeGroup ageGroup)
        {
            IDbConnection dbConnection = DButils.GetConnection(Properties);
            string query = "select p.participantid, p.name, p.age from participants p INNER JOIN participantstasks pt on p.participantid=pt.participantID INNER JOIN tasks t on pt.taskID1=t.taskid OR pt.taskID2 = t.taskid where t.type=@type AND t.agegroup=@agegroup";
            List<Participant> participants = new List<Participant>();

            using(var cmd = dbConnection.CreateCommand())
            {
                logger.InfoFormat("Searching participants with a given age and type of task");
                cmd.CommandText = query;
                
                var paramType = cmd.CreateParameter();
                paramType.ParameterName = "@type";
                paramType.Value = type.ToString();
                cmd.Parameters.Add(paramType);

                var paramAgeGroup = cmd.CreateParameter();
                paramAgeGroup.ParameterName = "@agegroup";
                paramAgeGroup.Value = ageGroup.ToString();
                cmd.Parameters.Add(paramAgeGroup);

                using(var dataR = cmd.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        var idP = dataR.GetInt32(0);
                        var name = dataR.GetString(1);
                        var age = dataR.GetInt32(2);
                        var participant = new Participant(idP, name, age);
                        participants.Add(participant);
                    }
                }
            }

            return participants;
        }

        public IEnumerable<Participant> FindAll()
        {
            IDbConnection dbConnection = DButils.GetConnection(Properties);
            IList<Participant> participants = new List<Participant>();
            
            using (var comm = dbConnection.CreateCommand())
            {
                logger.InfoFormat("all participants");
                comm.CommandText = "SELECT participantid, name, age FROM participants";
                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        var idP = dataR.GetInt32(0);
                        var name = dataR.GetString(1);
                        var age = dataR.GetInt32(2);
                        var participant = new Participant(idP, name, age);
                        participants.Add(participant);
                    }
                }
            }

            return participants;
        }

        public Participant Save(Participant e)
        {
            var dbConnection = DButils.GetConnection(Properties);

            using (var comm = dbConnection.CreateCommand())
            {
                logger.InfoFormat("saving participant {0}", e);
                comm.CommandText = "INSERT INTO participants VALUES (@participantid, @name, @age)";
                
                var paramID = comm.CreateParameter();
                paramID.ParameterName = "@participantid";
                paramID.Value = e.ID;
                comm.Parameters.Add(paramID);

                var paramName = comm.CreateParameter();
                paramName.ParameterName = "@name";
                paramName.Value = e.Name;
                comm.Parameters.Add(paramName);

                var paramAge = comm.CreateParameter();
                paramAge.ParameterName = "@age";
                paramAge.Value = e.Age;
                comm.Parameters.Add(paramAge);

                var result = comm.ExecuteNonQuery();
                if (result != 0) return e;
                logger.InfoFormat("Error DB");
                throw new Exception("No participant added!\n");
            }
        }

        public bool Delete(Participant e)
        {
            IDbConnection connection = DButils.GetConnection(Properties);
            using (var comm = connection.CreateCommand())
            {
                comm.CommandText = "DELETE FROM participants WHERE participantid=@participantid";
                IDbDataParameter paramID = comm.CreateParameter();
                paramID.ParameterName = "@participantid";
                paramID.Value = e.ID;
                comm.Parameters.Add(paramID);
                var dataR = comm.ExecuteNonQuery();
                if (dataR == 0)
                {
                    throw new Exception("can't delete participant");
                }
            }

            return false;
        }

        public Participant Update(Participant e)
        {
            IDbConnection connection = DButils.GetConnection(Properties);

            using (var comm = connection.CreateCommand())
            {
                logger.InfoFormat("updating participant {0}", e);
                comm.CommandText = "UPDATE participants SET name=@name, age=@age WHERE participantid=@participantid";
                
                var paramID = comm.CreateParameter();
                paramID.ParameterName = "@participantid";
                paramID.Value = e.ID;
                comm.Parameters.Add(paramID);

                var paramName = comm.CreateParameter();
                paramName.ParameterName = "@name";
                paramName.Value = e.Name;
                comm.Parameters.Add(paramName);

                var paramAge = comm.CreateParameter();
                paramAge.ParameterName = "@age";
                paramAge.Value = e.Age;
                comm.Parameters.Add(paramAge);

                var result = comm.ExecuteNonQuery();
                if (result != 0) return e;
                logger.InfoFormat("Error DB");
                throw new Exception("No participant updated!\n");
            }

            return e;
        }

        public int Size()
        {
            IDbConnection dbConnection = DButils.GetConnection(Properties);
            
            using (var comm = dbConnection.CreateCommand())
            {
                comm.CommandText = "SELECT count(*) FROM participants";
                using (var dataR = comm.ExecuteReader())
                {
                    if (!dataR.Read()) return 0;
                    logger.InfoFormat("count(*) participants");
                    return dataR.GetInt32(0);
                }
            }
        }
    }
}